### DISCUSSION TITLE
Why `Final` params are not allowed?

### DISCUSSION THREAD

**Discussion by Rudxain** [python/typing#1390] [Q&A]
python/peps#3102

**Comment by sobolevn** [comment]
> Why was that decision taken? What's the rationale?
> I haven't read 100% of that PEP, but I'm pretty sure my question isn't answered anywhere.
> I would really love to mark params as immutable! And this would make Py more suitable for purely-functional-programming

`Final` does not make objects immutable. It makes names un-reassignable. It might theoretically work like this:

```python
def some(a: Final[list[int]]):
   a.append(1)  # will still be ok
   a = []  # not ok, but why?
```

In my opinion, this was never implemented because this is not needed (prove me wrong by providing actual use-cases).
It is a common practice to reassign function params in a function body in Python.
And there's nothing wrong with that :)

**Reply by Rudxain** [reply]
I suppose you think `Final` is only useful in the global scope, because mutable global vars are bad practice in every language (specially if these are implicitly exported).

Idiomatic Python is (mostly) about being concise, so it makes sense to use `Final` sparingly.

I personally prefer to use `Final` everywhere possible, except when a var is in a narrow scope (short lifetime).

Functions can get big, so `Final` may be very useful in those cases. Of course, big fns should be split into modular fns, to prevent unmaintainable code from growing
